RecoHI/HiEvtPlaneAlgos/src/EvtPlaneProducer.cc

```diff
@@ -214,92 +365,92 @@ EvtPlaneProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+	  if(maxet_<0) maxet = maxTransverse[i];
+	  if(tower_energyet<minet) continue;
+	  if(tower_energyet>maxet) continue;
+	  if(EPDet[i]==HF) {
+	    double w = tower_energyet;
+	    if(loadDB_) w = tower_energyet*flat[i]->GetEtScale(vzr_sell,bin);
+	    if(EPOrder[i]==1 ) {
+	      if(MomConsWeight[i][0]=='y' && loadDB_ ) {
+		w = flat[i]->GetW(tower_energyet, vzr_sell, bin);
+	      }
+	      if(tower_eta<0 ) w=-w;
+	    }
+	    rp[i]->addParticle(w,tower_energyet,sin(EPOrder[i]*tower_phi),cos(EPOrder[i]*tower_phi),tower_eta);
+	  }
+	}
       } 
     }
+
+    //Castor part
+    
+ 
+    iEvent.getByToken(castorToken,castorCollection_);
+    
+    if(castorCollection_.isValid()){
+      for (std::vector<reco::CastorTower>::const_iterator j = castorCollection_->begin();j !=castorCollection_->end(); j++) {   
+       	tower_eta        = j->position().eta();
+	double labang = 2*TMath::ATan(TMath::Exp(-tower_eta));
+       	tower_phi        = j->position().phi();
+       	tower_energyet     = j->energy()*TMath::Sin(labang);
+	double minet = minet_;
+	double maxet = maxet_;
+       	for(int i = 0; i<NumEPNames; i++) {
+       	  if(EPDet[i]==Castor) {
+	    if(minet_<0) minet = minTransverse[i];
+	    if(maxet_<0) maxet = maxTransverse[i];
+	    if(tower_energyet<minet) continue;
+	    if(tower_energyet>maxet) continue;
+       	    double w = tower_energyet;
+       	    if(EPOrder[i]==1 ) {
+       	      if(MomConsWeight[i][0]=='y' && loadDB_ ) {
+       		w = flat[i]->GetW(tower_energyet, vzr_sell, bin);
+       	      }
+       	      if(tower_eta<0 ) w=-w;
+       	    }
+       	    rp[i]->addParticle(w,tower_energyet,sin(EPOrder[i]*tower_phi),cos(EPOrder[i]*tower_phi),tower_eta);
+       	  }
+       	}
+      } 
+    }
+
     //Tracking part
     
     double track_eta;
     double track_phi;
     double track_pt;
-
-    Handle<reco::TrackCollection> tracks;
-    iEvent.getByToken(trackCollection_, tracks);
-    
-    if(tracks.isValid()){
-      for(reco::TrackCollection::const_iterator j = tracks->begin(); j != tracks->end(); j++){
-	
-	//Find possible collections with command line: edmDumpEventContent *.root
-	
-	edm::Handle<reco::VertexCollection> vertex;
-	iEvent.getByToken(vtxCollection_, vertex);
-	
-	// find the vertex point and error
-
+   
+    iEvent.getByToken(trackToken, trackCollection_);
+    if(trackCollection_.isValid()){
+      for(reco::TrackCollection::const_iterator j = trackCollection_->begin(); j != trackCollection_->end(); j++){	
+	iEvent.getByLabel(vertexTag_, vertex_);	    
 	math::XYZPoint vtxPoint(0.0,0.0,0.0);
 	double vzErr =0.0, vxErr=0.0, vyErr=0.0;
-	if(vertex->size()>0) {
-	  vtxPoint=vertex->begin()->position();
-	  vzErr=vertex->begin()->zError();
-	  vxErr=vertex->begin()->xError();
-	  vyErr=vertex->begin()->yError();
+	if(vertex_->size()>0) {
+	  vtxPoint=vertex_->begin()->position();
+	  vzErr=vertex_->begin()->zError();
+	  vxErr=vertex_->begin()->xError();
+	  vyErr=vertex_->begin()->yError();
```

Comment:
the earlier comment about moving outside the track loop refers to these lines computing vtxPoint and  v*Err